### PR TITLE
[fixes] Add bug fixes from asan

### PR DESCRIPTION
**Pull request by zhoonit**
- [fixes] Add bug fixes from asan

```
**Changes proposed in this PR:**
- Fix issues from address sanitizer
- Enable deprecated-declaration warning but disable for some files

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: api/capi/src/nntrainer.cpp [hunk 1]
```diff
@@ -959,6 +959,11 @@ int ml_train_dataset_destroy(ml_train_dataset_h dataset) {
 
   check_feature_state();
 
+  if (dataset == nullptr) {
```
Comment (kparichay) [inline api/capi/src/nntrainer.cpp:962]: maybe I am missing something, but isn't this checked with nntrainer_internal.h L127? if not, then this check will be needed in all the destroy methods.

File: test/tizen_capi/unittest_tizen_capi.cpp [hunk 2]
```diff
@@ -19,6 +19,7 @@
  * @author      Jijoong Moon <jijoong.moon@samsung.com>
  * @bug         No known bugs
  */
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
```
Comment (kparichay) [inline test/tizen_capi/unittest_tizen_capi.cpp:22]: :100:

File: api/capi/src/nntrainer.cpp (hunk 1 above)
Comment (zhoonit) [reply api/capi/src/nntrainer.cpp]: You are correct, this is removed :)

**Review by kparichay — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Pull request merged**